include/aspect/introspection.h

```diff
@@ -57,17 +57,22 @@ namespace aspect
      */
     enum Type
     {
-      chemical_composition,
-      stress,
-      strain,
-      grain_size,
-      porosity,
-      density,
-      entropy,
-      generic,
-      unspecified
+      chemical_composition = 0,
+      stress = 1,
+      strain = 2,
+      grain_size = 3,
+      porosity = 4,
+      density = 5,
+      entropy = 6,
+      generic = 7,
+      unspecified = 8
     } type;
```

Comment:
Do you actually make use of these numeric values? My preference is to treat `enum`s as symbolic names independent of any numerical representation.

Reply:
Yes, I use these values to index into an std::vector below. That is why I explicitly specified them to make sure I can uniquely do the indexing. If I would just treat the `enum`s as symbolic names I would create a utility function with a `switch` statement to return an `unsigned int` depending on type. However, I think according to the standard I am actually allowed to implicitly cast the enums to the unsigned ints.

Reply:
I just dislike this from a conceptual perspective. It requires keeping track of things in different places.

What would you think of keeping the `enum` as just a list of symbolic names, and storing data instead of a fixed-size vector into a `std::map<enum_type, value_type>`?

Reply:
I am not happy about that solution because I do not know the performance implications. The cost of a lookup into a `std::map` already suprised us in the particle code before and the reason for this PR was to allow this lookup to be fast enough to be done multiple times per cell. What do you think about the alternative of keeping the `enum` as just a list of names, and having a helper function in an anonymous namespace in introspection.cc that is essentially:
```
unsigned int
index_of_type (enum type)
{
switch (type):
  case type_a:
    return 0;
  case type_b:
    return 1;
...
}
```
This way no code outside of Introspection can see that we translate the enum values into indices of a `std::vector`.

Reply:
But you still have to have two places that you keep in sync: The translation (whether in a function or by providing initializers) and the place where you store the *number* of `enum` elements.

I can be ok with the current state. Let's not make it more difficult!